lab_share_lib/processing/rabbit_message.py

```diff
@@ -1,41 +1,46 @@
+import logging
 from lab_share_lib.constants import (
     RABBITMQ_HEADER_KEY_ENCODER_TYPE,
     RABBITMQ_HEADER_KEY_SUBJECT,
     RABBITMQ_HEADER_KEY_VERSION,
     RABBITMQ_HEADER_VALUE_ENCODER_TYPE_DEFAULT,
 )
 
+LOGGER = logging.getLogger(__name__)
+
 
 class RabbitMessage:
     def __init__(self, headers, encoded_body):
         self.headers = headers
         self.encoded_body = encoded_body
 
-        self._subject = None
-        self._schema_version = None
         self._decoded_list = None
         self._message = None
 
     @property
     def encoder_type(self):
-        if RABBITMQ_HEADER_KEY_ENCODER_TYPE in self.headers.keys():
-            return self.headers[RABBITMQ_HEADER_KEY_ENCODER_TYPE]
-        return RABBITMQ_HEADER_VALUE_ENCODER_TYPE_DEFAULT
+        return self.headers.get(RABBITMQ_HEADER_KEY_ENCODER_TYPE, RABBITMQ_HEADER_VALUE_ENCODER_TYPE_DEFAULT)
 
     @property
     def subject(self):
-        if self._subject is None:
-            self._subject = self.headers[RABBITMQ_HEADER_KEY_SUBJECT]
-        return self._subject
+        return self.headers[RABBITMQ_HEADER_KEY_SUBJECT]
 
     @property
     def schema_version(self):
-        if self._schema_version is None:
-            self._schema_version = self.headers[RABBITMQ_HEADER_KEY_VERSION]
-        return self._schema_version
-
-    def decode(self, encoder):
-        self._decoded_list = list(encoder.decode(self.encoded_body, self.schema_version))
+        return self.headers[RABBITMQ_HEADER_KEY_VERSION]
+
+    def decode(self, possible_encoders):
+        exceptions = []
+        for encoder in possible_encoders:
+            try:
+                LOGGER.debug(f"Attempting to decode message with encoder class '{type(encoder).__name__}'.")
+                self._decoded_list = list(encoder.decode(self.encoded_body, self.schema_version))
```

Comment:
self._decoded_list contains all messages decoded by the first successful encoder from self._encoded_body. We still return the first message from the self._decoded_list by calling the message property. Although instances of this class support multiple messages in the body, it is intended to be used for single messages. Is that correct?

Reply:
That's right.  The format does not support more than one message.  It's for a single object only, and any file containing more than a single object would be invalid.  The `decode` method wraps that single object in a list when it returns so that the `message` property in this `RabbitMessage` class is still able to retrieve the only message from the list.